zaza/model.py

```diff
@@ -391,6 +392,66 @@ def get_app_ips(model_name, application_name):
 run_action_on_leader = sync_wrapper(async_run_action_on_leader)
 
 
+async def async_wait_for_application_states(model_name, states=None,
+                                            timeout=900):
+    """Wait for model to achieve the desired state
+
+    Check the workload status and workload status message for every unit of
+    every application. By default look for an 'active' workload status and a
+    message that start with one of the approved_message_prefixes.
+
+    Bespoke statuses and messages can be passed in with states. states takes
+    the form:
+
+    {
+        'app': {
+            'workload-status': 'blocked',
+            'workload-status-message': 'No requests without a prod'}
+        'anotherapp': {
+            'workload-status-message': 'Unit is super ready'}}
+
+
+    :param model_name: Name of model to query.
+    :type model_name: str
+    :param states: Stest to look for
+    :type states: dict
+    :param timeout: Time to wait for status to be achieved
+    :type timeout: int
+    """
+    approved_message_prefixes = ('ready', 'Ready', 'Unit is ready')
+
+    if not states:
+        states = {}
+    async with run_in_model(model_name) as model:
+        logging.info("Waiting for all units to be idle")
+        await model.block_until(
+            lambda: model.all_units_idle(), timeout=timeout)
+        for application in model.applications:
+            check_info = states.get(application, {})
+            for unit in model.applications[application].units:
+                logging.info("Checking workload status of {}".format(
+                    unit.entity_id))
+                await model.block_until(
```

Comment:
Not sure how, but if we could stick a check for 'error' status in here so we bail early that would be great. This is not a blocker for this first round.

Reply:
I have added checks for error state

Reply:
I have added checks for error state